File: instrumentation/jetty-httpclient/jetty-httpclient-9.2/library/src/main/java/io/opentelemetry/instrumentation/jetty/httpclient/v9_2/internal/Experimental.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.jetty.httpclient.v9_2.internal;

import static java.util.logging.Level.FINE;

import io.opentelemetry.instrumentation.jetty.httpclient.v9_2.JettyClientTelemetryBuilder;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.logging.Logger;
import javax.annotation.Nullable;

/**
 * This class is internal and experimental. Its APIs are unstable and can change at any time. Its
 * APIs (or a version of them) may be promoted to the public stable API in the future, but no
 * guarantees are made.
 */
public class Experimental {

  private static final Logger logger = Logger.getLogger(Experimental.class.getName());

  @Nullable
  private static final Method emitExperimentalTelemetryMethod =
      getEmitExperimentalTelemetryMethod();

  public void setEmitExperimentalTelemetry(
      JettyClientTelemetryBuilder builder, boolean emitExperimentalTelemetry) {

    if (emitExperimentalTelemetryMethod != null) {
      try {
        emitExperimentalTelemetryMethod.invoke(builder, emitExperimentalTelemetry);
      } catch (IllegalAccessException | InvocationTargetException e) {
        logger.log(FINE, e.getMessage(), e);
      }
    }
  }

  @Nullable
  private static Method getEmitExperimentalTelemetryMethod() {
    try {
      Method method =
          JettyClientTelemetryBuilder.class.getDeclaredMethod(
              "setEmitExperimentalHttpClientMetrics", boolean.class);
      method.setAccessible(true);
      return method;
    } catch (NoSuchMethodException e) {
      logger.log(FINE, e.getMessage(), e);
      return null;
    }
  }
}

File: instrumentation/ktor/ktor-2-common/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/common/client/internal/Experimental.kt
package io.opentelemetry.instrumentation.ktor.v2_0.common.client.internal

import io.opentelemetry.instrumentation.ktor.v2_0.common.client.AbstractKtorClientTelemetryBuilder
import java.util.function.BiConsumer

class Experimental private constructor() {

    companion object {
        private var setEmitExperimentalTelemetry: BiConsumer<AbstractKtorClientTelemetryBuilder, Boolean>? = null

        fun emitExperimentalTelemetry(
            builder: AbstractKtorClientTelemetryBuilder
        ) {
            if (setEmitExperimentalTelemetry != null) {
                setEmitExperimentalTelemetry!!.accept(builder, true)
            }
        }

        fun setSetEmitExperimentalTelemetry(
            setEmitExperimentalTelemetry: BiConsumer<AbstractKtorClientTelemetryBuilder, Boolean>?
        ) {
            Companion.setEmitExperimentalTelemetry = setEmitExperimentalTelemetry
        }
    }
}
